contracts/gov/proposals/TokenSaleProposal.sol

```diff
@@ -0,0 +1,381 @@
+                (tierView.minAllocationPerUser <= saleTokenAmount &&
+                    saleTokenAmount <= tierView.maxAllocationPerUser),
+            "TSP: wrong allocation"
+        );
+
+        require(
+            tierInfo.totalSold + saleTokenAmount <= tierView.totalTokenProvided,
+            "TSP: insufficient sale token amount"
+        );
+
+        return saleTokenAmount;
+    }
+
+    function getVestingWithdrawAmounts(
+        uint256[] calldata tierIds
+    ) public view returns (uint256[] memory vestingWithdrawAmounts) {
+        vestingWithdrawAmounts = new uint256[](tierIds.length);
+
+        for (uint256 i = 0; i < tierIds.length; i++) {
+            vestingWithdrawAmounts[i] = _getVestingWithdrawAmount(tierIds[i]);
+        }
+    }
+
+    function getRecoverAmounts(
+        uint256[] calldata tierIds
+    ) public view returns (uint256[] memory recoveringAmounts) {
+        recoveringAmounts = new uint256[](tierIds.length);
+
+        for (uint256 i = 0; i < recoveringAmounts.length; i++) {
+            recoveringAmounts[i] = _getRecoverAmount(tierIds[i]);
+        }
+    }
+
+    function getTiers(
+        uint256 offset,
+        uint256 limit
+    ) external view returns (TierView[] memory tierViews) {
+        uint256 to = (offset + limit).min(latestTierId).max(offset);
+
+        tierViews = new TierView[](to - offset);
+
+        for (uint256 i = offset; i < to; i++) {
+            tierViews[i - offset] = _tiers[i + 1].tierView;
+        }
+    }
+
+    function _createTier(TierView calldata tierView) internal {
+        require(tierView.saleTokenAddress != address(0), "TSP: sale token cannot be zero");
+        require(tierView.saleTokenAddress != ETHEREUM_ADDRESS, "TSP: cannot sale native currency");
+
+        require(tierView.totalTokenProvided != 0, "TSP: sale token is not provided");
+
+        require(
+            tierView.saleStartTime <= tierView.saleEndTime,
+            "TSP: saleEndTime is less than saleStartTime"
+        );
+
+        require(
+            tierView.minAllocationPerUser <= tierView.maxAllocationPerUser,
+            "TSP: wrong allocation"
+        );
+
+        require(tierView.vestingPercentage <= PERCENTAGE_100, "TSP: vestingPercentage > 100%");
+
+        require(tierView.vestingSettings.unlockStep != 0, "TSP: unlockStep cannot be zero");
+        require(
+            tierView.vestingSettings.vestingDuration >= tierView.vestingSettings.unlockStep,
+            "TSP: vestingDuration should greater than unlock step"
+        );
+
+        _amountsToSell[tierView.saleTokenAddress] += tierView.totalTokenProvided;
+
+        require(
+            IERC20(tierView.saleTokenAddress).balanceOf(address(this)) >=
+                _amountsToSell[tierView.saleTokenAddress],
+            "TSP: insufficient TSP balance"
+        );
+
+        Tier storage tier = _tiers[++latestTierId];
+
+        tier.tierView = tierView;
+
+        TierInfo storage tierInfo = tier.tierInfo;
```

Comment:
```suggestion
        TierInfo storage tierInfo = tier.tierInfo;
        
        tier.tierView = tierView;
```

Reply:
Let's assign the `tierView` variable at the end of the execution when all the checks are passed (gas savings)

Reply:
done